brian2genn/templates/engine.cpp

```diff
@@ -156,24 +213,45 @@ void engine::run(double duration, //!< Duration of time to run the model for
           _run_{{spkGen.name}}_codeobject();
           push{{spkGen.name}}SpikesToDevice();
           {% endfor %}
+          {% set spikes_pulled = [] %}
           {% for spkMon in spike_monitor_models %}
           {% if (spkMon.notSpikeGeneratorGroup) %}
+          {% if not spkMon.neuronGroup in spikes_pulled %}
           pull{{spkMon.neuronGroup}}SpikesFromDevice();
+          {% if spikes_pulled.append(spkMon.neuronGroup) %}{% endif %}
+          {% endif %}
           {% endif %}
           {% endfor %}
           {% for rateMon in rate_monitor_models %}
           {% if (rateMon.notSpikeGeneratorGroup) %}
+          {% if not rateMon.neuronGroup in spikes_pulled %}
           pull{{rateMon.neuronGroup}}SpikesFromDevice();
+          {% if spikes_pulled.append(rateMon.neuronGroup) %}{% endif %}
+          {% endif %}
           {% endif %}
           {% endfor %}
+          {% set states_pulled = [] %}
           {% for sm in state_monitor_models %}
+          {% if not sm.monitored in states_pulled %}
           pull{{sm.monitored}}StateFromDevice();
+          {% if states_pulled.append(sm.monitored) %}{% endif %}
+          {% endif %}
+          {% endfor %}
+          {% for run_reg in run_regularly_operations %}
+            {% for var in run_reg['read'] %}
+            {% if not var in ['t', 'dt'] %}
+            {% if not run_reg['owner'].variables[var].owner.name in states_pulled %}
+            pull{{run_reg['owner'].variables[var].owner.name}}StateFromDevice();
+            {% if states_pulled.append(run_reg['owner'].variables[var].owner.name) %}{% endif %}
+            {% endif %}
+            {% endif %}
+            {% endfor %}
           {% endfor %}
```

Comment:
doesn't this also mean that you are pulling variables every timestep, regardless whether the run_regularly operations will take place in the next iteration? I wonder whether one should not test here whether the variable will be needed?

Reply:
Right, actually it would make much more sense to put the `pullStateFromDevice` together with the actual execution of the operation. I think I got confused by all the state pulling that is going on here for the monitors, which is preparing things for the next time step as well.

Reply:
Looking at the code I realized that while it is not intuitive to have the state pulling at the end of the time step, it allows us to not do it twice if the state has already been pulled for a state monitor. I'll leave it this way then and just add a check that the next time step will execute it.